### ISSUE TITLE
--experimental_guard_against_concurrent_changes is fragile.

### ISSUE THREAD

**Issue by buchgr** [bazelbuild/bazel#5566]
On Buildkite we are always seeing warnings like the below:

```
WARNING: /private/var/tmp/_bazel_buildkite/4152b1dabe54c24162562a836b8cb156/execroot/io_bazel/third_party/java/jdk/langtools/BUILD was modified during execution
```

**Comment by buchgr** [comment]
We mostly see it on macOS.

**Comment by sha1n** [comment]
@buchgr I noticed these messages started popping occasionally on our macOS builds at Wix (not sure about Linux) as well, but I have only seen them referring to jar files if it adds any information. 

One thing I noticed is that they only started showing after we increased `--jobs` beyond the number of available CPUs.
We are currently experimenting with shared remote-caching for developers and use the following flags as a default after experimenting with several different resource allocation settings:

```
build --local_cpu_resources=HOST_CPUS
build --jobs=HOST_CPUS*2
```
As far as I understand how it works, this only means that those files will not be uploaded to the remote-cache and I do not have to worry about consistency. Is that correct?
